e2e/src/bin/historic_trades.rs

```diff
@@ -0,0 +1,304 @@
+    match meta.settled_xrate {
+        Some(xrate) if xrate > meta.limit_price => TradeResult::OverlyPessimistic,
+        _ => TradeResult::UnreasonableOrderNotMatched,
+    }
+}
+
+struct OrderMetadata {
+    effective_sell_amount: f64,
+    limit_price: f64,
+    estimated_limit_price: Option<f64>,
+    settled_xrate: Option<f64>,
+    trade: Option<Trade>,
+}
+
+impl OrderMetadata {
+    /// Computes order metadata based on a batch settlement (solution), order
+    /// data and a `Pricegraph` instance.
+    fn compute(settlement: Option<&Settlement>, order: &Element, pricegraph: &Pricegraph) -> Self {
+        let effective_sell_amount =
+            pricegraph::num::u256_to_f64(order.balance).min(order.remaining_sell_amount as _);
+        let limit_price = order.price.numerator as f64 / order.price.denominator as f64;
+        let estimated_limit_price =
+            pricegraph.estimate_limit_price(order.pair, effective_sell_amount);
+        let settled_xrate =
+            settlement.and_then(|settlement| find_settled_exchange_rate(settlement, order.pair));
+        let trade = settlement.and_then(|settlement| find_trade(settlement, order));
+
+        OrderMetadata {
+            effective_sell_amount,
+            limit_price,
+            estimated_limit_price,
+            settled_xrate,
+            trade,
+        }
+    }
+
+    /// Returns `true` if an order is considered reasonably priced, `false`
+    /// otherwise. An order is considered reasonably priced if its limit price
+    /// is lower (or "worse" for the order owner) than the `Pricegraph`
+    /// estimated limit price for the order's token pair.
+    fn is_reasonably_priced_order(&self) -> bool {
+        match self.estimated_limit_price {
+            Some(estimated_limit_price) => self.limit_price <= estimated_limit_price,
+            None => false,
+        }
+    }
+
+    /// The ratio at which the order is filled - `0` being not filled at all and
+    /// `1` if the order is fully filled.
+    ///
+    /// Returns `None` if the order was not traded.
+    fn fill_ratio(&self) -> Option<f64> {
+        self.trade
+            .as_ref()
+            .map(|trade| trade.executed_sell_amount as f64 / self.effective_sell_amount)
+    }
+}
+
+fn find_trade(settlement: &Settlement, order: &Element) -> Option<Trade> {
+    settlement
+        .trades
+        .iter()
+        .find(|trade| (trade.owner, trade.order_id) == (order.user, order.id))
+        .cloned()
+}
+
+fn find_settled_exchange_rate(settlement: &Settlement, pair: TokenPair) -> Option<f64> {
+    let buy_token_index = settlement
+        .solution
+        .token_ids_for_price
+        .iter()
+        .position(|&id| id == pair.buy)?;
+    let sell_token_index = settlement
+        .solution
+        .token_ids_for_price
+        .iter()
+        .position(|&id| id == pair.sell)?;
+
+    // NOTE: This is unintuitive, but the echange rate is computed as the sell
```

Comment:
Maybe we can also argue with the "unit coherence" argument.

Prices p(x) are denoted in OWL "per" token (OWL/t_x)
The exchange rate we are looking for is expressed as t_buy per t_sell (t_buy/t_sell).

Therefore p(sell)/p(buy) = (OWL/t_sell) / (OWL/t_buy) = (OWL/t_sell) * (t_buy/OWL) = t_buy/t_sell